### PR TITLE
Fix for Issue #1278 : Created debugging file

### LINKED ISSUE #1278: Add docs on how to debug the backend
We have a bunch of VSCode configs (e.g., [`Launch Telescope`](https://github.com/Seneca-CDOT/telescope/blob/master/.vscode/launch.json#L4-L11)) that allow devs to easily debug the backend from within VSCode.  However, if you haven't used them before, it might not be obvious.

Let's add some docs on how to do this.  See https://code.visualstudio.com/docs/editor/debugging#_launch-configurations for some info.

Let's cover how to start the server from VSCode, how to see the console for errors, how to set breakpoints, etc.

### PR DESCRIPTION
**Pull request by mamadou-diallo**
This is to resolve issue #1278

## Type of Change

- [x] **Documentation Update**: Change which improves documentation

## Description

Added documentation that teaches one that isn't too familiar with debugging on VS. 
## Checklist

- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [x] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Review by humphd — CHANGES_REQUESTED** [review]
This is failing CI due to formatting in the markdown file:

```
> prettier --check "./**/*.{md,jsx,json,html,css,js,yml}"

Checking formatting...
[warn] docs/debugging.md
[warn] Code style issues found in the above file(s). Forgot to run Prettier?
```

Can you please run `npm install` on your machine and then `npm run prettier` and add the changed file, re-commit, and push on this branch to add to the PR.

**Comment by mamadou-diallo** [comment]
@humphd As requested I've made the required changes :)

File: docs/debugging.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+### Setting up VS Code and Launching Telescope :
+
+To begin debugging one must retrieve the launch.json file.
+This file contains **Important** debugging configuration information which should be located in the .vscode folder for telescope to run.
+
+Once the launch.json file has been located, you can move on to the "Run View" tab.  
+[VS Run View Screenshot](https://drive.google.com/file/d/1dvXhen8_eT68qEMa8Z6OpejDiMOG1pLN)
+Visual Studio Code will automatically try to detect the debug environment but in the case that it fails you must select "Node.js".
+[VS Debug Environment Screenshot](https://drive.google.com/file/d/1pvL4RJRJGNLb1LmDZRN3x5UCusxGHdJU/view?usp=sharing)
```
Comment (humphd) [inline docs/debugging.md:16]: Let's add the image to the repo under [`docs/images/`](https://github.com/Seneca-CDOT/telescope/tree/master/docs/images)

File: docs/debugging.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+## Debugging Tutorial
+
+### For debugging purposes the following must be installed :
+
+[Visual Studio Code](https://code.visualstudio.com/)
+[Node.js](https://www.npmjs.com/) - to setup node.js you may follow the instructions in [Environment Setup](https://github.com/Seneca-CDOT/telescope/blob/master/docs/environment-setup.md)
+
+### Setting up VS Code and Launching Telescope :
+
+To begin debugging one must retrieve the launch.json file.
```
Comment (humphd) [inline docs/debugging.md:10]: I'm not sure what this means.  We already have a `launch.json` file in the repo.

File: docs/debugging.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+This file contains **Important** debugging configuration information which should be located in the .vscode folder for telescope to run.
```
Comment (humphd) [inline docs/debugging.md:11]: `important` (no capital I)

File: docs/debugging.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+Once the launch.json file has been located, you can move on to the "Run View" tab.  
+[VS Run View Screenshot](https://drive.google.com/file/d/1dvXhen8_eT68qEMa8Z6OpejDiMOG1pLN)
+Visual Studio Code will automatically try to detect the debug environment but in the case that it fails you must select "Node.js".
+[VS Debug Environment Screenshot](https://drive.google.com/file/d/1pvL4RJRJGNLb1LmDZRN3x5UCusxGHdJU/view?usp=sharing)
+To start the server you must now click on the "Launch Icon" this will run the server.
+
+[VS Launch Telescope Screenshot](https://drive.google.com/file/d/1dvXhen8_eT68qEMa8Z6OpejDiMOG1pLN)
+
+At the bottom left you will see your Call Stack and if successfully ran, index.js and master.js should be up and running.
+[VS Call Stack Screenshot](https://drive.google.com/file/d/1trLs89HHFisIy4Z0LqlMuK7LyUYD3zh7/view?usp=sharing)
```
Comment (humphd) [inline docs/debugging.md:22]: Same deal here, import image to the repo.

File: docs/debugging.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+### Debugging :
+
+Now that we know how to launch the server, we may now proceed to debugging.
+In this case we will use index.js as an example. When debugging you can set breakpoints(Using F9 or go through toolbar menu Run->Toggle Breakpoint) that will appear as a red dot on the left side of a line. With breakpoints in place, you may now use the Debug toolbar to toggle different debugging options.
```
Comment (humphd) [inline docs/debugging.md:27]: space before (

File: docs/debugging.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+[VS Breakpoint Screenshot](https://drive.google.com/file/d/1H4uGT2GZklNkXt0a5duAhswK6TgkX9k3)
```
Comment (humphd) [inline docs/debugging.md:28]: Import image to repo

File: docs/debugging.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+Once breakpoints are set and the server is running, you may now view the current state of the system and view variables, functions that are being called, etc.
+
+[VS Variables + Problems Screenshot](https://drive.google.com/drive/u/0/folders/1d_Om0IY7n-clKxJR_aXH9ZM1oJtZFYRc)
```
Comment (humphd) [inline docs/debugging.md:32]: Import image to repo

File: docs/debugging.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+At the bottom you may also see errors ( recognized as problems ).
```
Comment (humphd) [inline docs/debugging.md:34]: remove spaces after and before (...)

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by mamadou-diallo** [comment]
@humphd All the changes you've requested have been made. I've added a bit on content for the "Before you begin" to make it a bit more intuitive for a user. 
Sorry for the delay, notifications never got pushed out to me as I was flying out of the province.

File: docs/debugging.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
+## Debugging Tutorial
+
+### For debugging purposes the following must be installed :
+
+[Visual Studio Code](https://code.visualstudio.com/)
+[Node.js](https://www.npmjs.com/) - to setup node.js you may follow the instructions in [Environment Setup](https://github.com/Seneca-CDOT/telescope/blob/master/docs/environment-setup.md)
+
+### Setting up VS Code and Launching Telescope :
+
+To begin debugging one must make sure that the telescope launch.json file can be found as VS code will automatically generate it's own launch.json file if none is found.
```
Comment (humphd) [inline docs/debugging.md:10]: This sentence is true, but we already have it in our repo, so it's not necessary.

File: docs/debugging.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+This launch.json file contains **important** debugging configuration information which should be located in the .vscode folder for telescope to run.
+
+Once the launch.json file has been located, you can move on to the "Run View" tab.
```
Comment (humphd) [inline docs/debugging.md:13]: Start your discussion here re: Run View tab.

File: docs/debugging.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+[Visual Studio Code](https://code.visualstudio.com/)
+[Node.js](https://www.npmjs.com/) - to setup node.js you may follow the instructions in [Environment Setup](https://github.com/Seneca-CDOT/telescope/blob/master/docs/environment-setup.md)
+
+### Setting up VS Code and Launching Telescope :
+
+To begin debugging one must make sure that the telescope launch.json file can be found as VS code will automatically generate it's own launch.json file if none is found.
+This launch.json file contains **important** debugging configuration information which should be located in the .vscode folder for telescope to run.
+
+Once the launch.json file has been located, you can move on to the "Run View" tab.  
+[VS Run View Screenshot](https://drive.google.com/file/d/1dvXhen8_eT68qEMa8Z6OpejDiMOG1pLN)
```
Comment (humphd) [inline docs/debugging.md:14]: These images are still pointing to Google Drive, please include in the repo and use relative links.

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by mamadou-diallo** [comment]
@humphd I've done all the requested changes ! Should I be closing all these conversations?

**Review by humphd — DISMISSED** [review]

File: docs/debugging.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+[Visual Studio Code](https://code.visualstudio.com/)
+[Node.js](https://www.npmjs.com/) - to setup node.js you may follow the instructions in [Environment Setup](https://github.com/Seneca-CDOT/telescope/blob/master/docs/environment-setup.md)
+
+### Setting up VS Code and Launching Telescope :
+
+To begin debugging one must select the "Run View" tab. This will display a menu that for debugging purposes.
+
+![VS Run View Screenshot](./images/VS-Run-View-Screenshot.png)
+Visual Studio Code will automatically try to detect the debug environment but in the case that it fails you must select "Node.js".
```
Comment (manekenpix) [inline docs/debugging.md:13]: I think this needs an extra newline:

image

**Review by manekenpix — CHANGES_REQUESTED** [review]
Looking good, just added one comment.
Also, I think it'd be interesting to include all the different debugging options that Telescope has:

image

File: docs/debugging.md (hunk 12 above)
Comment (mamadou-diallo) [reply docs/debugging.md]: @manekenpix I've fixed that new line, I also added an additional screenshot to display all the features that can be utilized with VS code.

I now see that the code is failing due to a new feature that's been added... If you give me a hint ( as I don't see anything in the docs )  on how to resolve this I'll be more than happy to fix it ! 

Thanks !

Comment (manekenpix) [reply docs/debugging.md]: @mamadou-diallo I think rebasing on upstream master will fix the issue

Comment (manekenpix) [reply docs/debugging.md]: @mamadou-diallo also, I saw that you added VScode features, but I meant telescope's debugging options. If you take a look at the screenshot I posted in my first comment, you can see how you can choose a debugging option depending on what you want to do:
- Launch Telescope (for debugging the Telescope app)
- Launch Auto Deployment (for debugging the autodeployment server)
- Launch All Tests (for debugging our test set)
- Launch Opened Test File (for debugging an opened test in VScode)

For detailed info about these options, check out our [launch,json](https://github.com/Seneca-CDOT/telescope/blob/master/.vscode/launch.json) in the `.vscode` folder.

I think it'd be more interesting if the document this PR is adding focused on the 4 different debugging options that Telescope has (which is what's really helpful for developers) more than explaining VScode features.

Let me know what you think.

Comment (mamadou-diallo) [reply docs/debugging.md]: @manekenpix Oops ! I didn't get what you meant initially but I don't mind doing that correction.

**Comment by mamadou-diallo** [comment]
@manekenpix @humphd I think I did a mistake here trying to reset my branch to telescope/master and force-pushed my code to telescope ... ( ultimately I managed to do what I was looking to do here & created a new PR #1517 ) sorry in advance.

**Pull request closed without merging**